usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2770,26 +2788,78 @@ static CK_RV ccatok_pkey_update(STDLL_TokData_t *tokdata, OBJECT *key_obj,
         goto done;
     }
 
+retry:
     /* Transform the secure key into a protected key */
     ret = ccatok_pkey_skey2pkey(tokdata, skey_attr, &pkey_attr, aes_xts);
     if (ret != CKR_OK) {
         TRACE_ERROR("protected key creation failed with rc=0x%lx\n",ret);
         goto done;
     }
 
-    /* Check if the new pkey's verification pattern matches the one in
-     * cca_data. This should always be the case, because we just
-     * created the pkey with the current MK. */
+    if (pthread_rwlock_rdlock(&cca_data->pkey_rwlock)) {
+        TRACE_ERROR("%s Failed to lock pkey lock\n", __func__);
+        ret = CKR_CANT_LOCK;
+        goto done;
+    }
+
+    /*
+     * Check if the new pkey's verification pattern matches the one in
+     * cca_data. This should always be the case, except there was a live
+     * guest relocation (LGR) in the middle of the process.
+     */
     vp_offset = pkey_attr->ulValueLen - PKEY_MK_VP_LENGTH;
     if (memcmp(&cca_data->pkey_mk_vp, (CK_BYTE *)pkey_attr->pValue + vp_offset,
                PKEY_MK_VP_LENGTH) != 0) {
-        TRACE_ERROR("vp of this pkey does not match with the one in cca_data (should not occur)\n");
+        TRACE_ERROR("vp of this pkey does not match with the one in cca_data\n");
         ret = CKR_FUNCTION_FAILED;
-        goto done;
     }
 
-    /* Now update the key obj. If it's a token obj, it will be also updated
-     * in the repository. */
+    if (aes_xts) {
+        /*
+         * Check if the new pkey's verification pattern matches the one in
+         * cca_data. This should always be the case, except there was a live
+         * guest relocation (LGR) in the middle of the process.
+         * AES XTS has two keys, two keys are concatenated.
+         * Second key is checked above and the first key is checked here
+         */
+        vp_offset = pkey_attr->ulValueLen / 2 - PKEY_MK_VP_LENGTH;
+        if (memcmp(&cca_data->pkey_mk_vp, (CK_BYTE *)pkey_attr->pValue + vp_offset,
+                   PKEY_MK_VP_LENGTH) != 0) {
+            TRACE_ERROR("vp of this pkey does not match with the one in cca_data\n");
+            ret = CKR_FUNCTION_FAILED;
+        }
+    }
+
+    if (pthread_rwlock_unlock(&cca_data->pkey_rwlock))
+        TRACE_ERROR("%s Failed to unlock pkey lock\n", __func__);
```

Comment:
Please use { ... } around the TRACE_ERROR stmt. Its a single line block, so { } would not be needed, but TRACE_ERROR is a macro and you don't know to what it expands. If it expands to multiple lines, then only the very first expanded line would be part of the if stmt, but the remaining lines not.

Reply:
fixed.